aws/resource_aws_cloudhsm2_cluster.go

```diff
@@ -0,0 +1,384 @@
+
+	subnetIdsSet := d.Get("subnet_ids").(*schema.Set)
+	subnetIds := make([]*string, subnetIdsSet.Len())
+	for i, subnetId := range subnetIdsSet.List() {
+		subnetIds[i] = aws.String(subnetId.(string))
+	}
+
+	input := &cloudhsmv2.CreateClusterInput{
+		HsmType:   aws.String(d.Get("hsm_type").(string)),
+		SubnetIds: subnetIds,
+	}
+
+	backupId := d.Get("backup_identifier").(string)
+	if len(backupId) != 0 {
+		input.SourceBackupId = aws.String(backupId)
+	}
+
+	log.Printf("[DEBUG] CloudHSMv2 Cluster create %s", input)
+
+	var output *cloudhsmv2.CreateClusterOutput
+
+	err := resource.Retry(180*time.Second, func() *resource.RetryError {
+		var err error
+		output, err = cloudhsm2.CreateCluster(input)
+		if err != nil {
+			if isAWSErr(err, cloudhsmv2.ErrCodeCloudHsmInternalFailureException, "request was rejected because of an AWS CloudHSM internal failure") {
+				log.Printf("[DEBUG] CloudHSMv2 Cluster re-try creating %s", input)
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	d.SetId(aws.StringValue(output.Cluster.ClusterId))
+	log.Printf("[INFO] CloudHSMv2 Cluster ID: %s", d.Id())
+	log.Println("[INFO] Waiting for CloudHSMv2 Cluster to be available")
+
+	targetState := cloudhsmv2.ClusterStateUninitialized
+	if len(backupId) > 0 {
+		targetState = cloudhsmv2.ClusterStateActive
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{cloudhsmv2.ClusterStateCreateInProgress, cloudhsmv2.ClusterStateInitializeInProgress},
+		Target:     []string{targetState},
+		Refresh:    resourceAwsCloudHsm2ClusterRefreshFunc(d, meta),
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		MinTimeout: 30 * time.Second,
+		Delay:      30 * time.Second,
+	}
+
+	// Wait, catching any errors
+	_, errWait := stateConf.WaitForState()
+	if errWait != nil {
+		if len(backupId) == 0 {
+			return fmt.Errorf("[WARN] Error waiting for CloudHSMv2 Cluster state to be \"UNINITIALIZED\": %s", errWait)
+		} else {
+			return fmt.Errorf("[WARN] Error waiting for CloudHSMv2 Cluster state to be \"ACTIVE\": %s", errWait)
+		}
+	}
+
+	if err := setTagsAwsCloudHsm2Cluster(cloudhsm2, d); err != nil {
+		return err
+	}
+
+	return resourceAwsCloudHsm2ClusterRead(d, meta)
+}
+
+func resourceAwsCloudHsm2ClusterRead(d *schema.ResourceData, meta interface{}) error {
+
+	cluster, err := describeCloudHsm2Cluster(d.Id(), meta)
+
+	if cluster == nil {
+		log.Printf("[WARN] CloudHSMv2 Cluster (%s) not found", d.Id())
+		d.SetId("")
+		return err
+	}
+
+	log.Printf("[INFO] Reading CloudHSMv2 Cluster Information: %s", d.Id())
+
+	d.Set("cluster_id", cluster.ClusterId)
+	d.Set("cluster_state", cluster.State)
+	d.Set("security_group_id", cluster.SecurityGroup)
+	d.Set("vpc_id", cluster.VpcId)
+	d.Set("backup_identifier", cluster.SourceBackupId)
+	d.Set("hsm_type", cluster.HsmType)
+	d.Set("cluster_certificate", readCloudHsm2ClusterCertificates(cluster))
```

Comment:
When using `d.Set()` with aggregate types (TypeList, TypeSet, TypeMap), [we should perform error checking to prevent issues where the code is not properly able to set the Terraform state](https://www.terraform.io/docs/extend/best-practices/detecting-drift.html#error-checking-aggregate-types). e.g.

```go
if err := d.Set("cluster_certificates", readCloudHsm2ClusterCertificates(cluster)); err != nil {
  return fmt.Errorf("error setting cluster_certificates: %s", err)
}
```

Note the above example also fixes the typo of a missing `s` in the attribute name.